LibWeb: Invalidate the display list when calling `set_needs_display()`

tcl3 opened the pull request:
Previously, content wasn't always redrawn when a repaint was necessary. This PR fixes the following issues

* The media now repainted when its state changes
* The cursor is redrawn when moved
* Animated images are repainted when their animation state changes

Example from: https://en.wikipedia.org/wiki/File:United_States_Navy_Band_-_Amhr%C3%A1n_na_bhFiann.ogg
NOTE: I had to disable scripts to get this page to work correctly.

Before:

https://github.com/user-attachments/assets/e532c8c0-5737-42a0-8792-1bc8c0e25fb3

After:


https://github.com/user-attachments/assets/bd9fc5ea-7153-46f4-9323-51c4f94b5dea

Fixes: #1256

trflynn89 commented:
Oh man...I think this is just one of several locations that need to be updated in the media element classes. Playing videos and interacting with the controls (e.g. volume) also update very slowly, if at all.

I'm running with this diff locally, and the media element behaves better for audio and video:
```diff
diff --git a/Userland/Libraries/LibWeb/HTML/AudioTrack.cpp b/Userland/Libraries/LibWeb/HTML/AudioTrack.cpp
index 72dd57d543..40a3562f88 100644
--- a/Userland/Libraries/LibWeb/HTML/AudioTrack.cpp
+++ b/Userland/Libraries/LibWeb/HTML/AudioTrack.cpp
@@ -31,8 +31,10 @@ AudioTrack::AudioTrack(JS::Realm& realm, JS::NonnullGCPtr<HTMLMediaElement> medi
     , m_audio_plugin(Platform::AudioCodecPlugin::create(move(loader)).release_value_but_fixme_should_propagate_errors())
 {
     m_audio_plugin->on_playback_position_updated = [this](auto position) {
-        if (auto* paintable = m_media_element->paintable())
+        if (auto* paintable = m_media_element->paintable()) {
             paintable->set_needs_display();
+            paintable->document().invalidate_display_list();
+        }

         auto playback_position = static_cast<double>(position.to_milliseconds()) / 1000.0;
         m_media_element->set_current_playback_position(playback_position);
diff --git a/Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp b/Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
index 72aa8bda17..21442ea31d 100644
--- a/Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
+++ b/Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
@@ -336,8 +336,10 @@ void HTMLMediaElement::set_duration(double duration)

     m_duration = duration;

-    if (auto* paintable = this->paintable())
+    if (auto* paintable = this->paintable()) {
         paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }
 }

 WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::Promise>> HTMLMediaElement::play()
@@ -432,8 +434,10 @@ void HTMLMediaElement::volume_or_muted_attribute_changed()

     // FIXME: Then, if the media element is not allowed to play, the user agent must run the internal pause steps for the media element.

-    if (auto* paintable = this->paintable())
+    if (auto* paintable = this->paintable()) {
         paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }

     on_volume_change();
 }
@@ -1636,8 +1640,10 @@ void HTMLMediaElement::set_show_poster(bool show_poster)

     m_show_poster = show_poster;

-    if (auto* paintable = this->paintable())
+    if (auto* paintable = this->paintable()) {
         paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }
 }

 void HTMLMediaElement::set_paused(bool paused)
@@ -1654,8 +1660,11 @@ void HTMLMediaElement::set_paused(bool paused)
             document().page().client().page_did_change_audio_play_state(AudioPlayState::Paused);
     }

-    if (auto* paintable = this->paintable())
+    if (auto* paintable = this->paintable()) {
         paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }
+
     set_needs_style_update(true);
 }

@@ -1983,8 +1992,10 @@ void HTMLMediaElement::set_layout_display_time(Badge<Painting::MediaPaintable>,

     m_display_time = move(display_time);

-    if (auto* paintable = this->paintable())
+    if (auto* paintable = this->paintable()) {
         paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }
 }

 double HTMLMediaElement::layout_display_time(Badge<Painting::MediaPaintable>) const
diff --git a/Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp b/Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp
index 9d445ecdd7..c00875a211 100644
--- a/Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp
+++ b/Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp
@@ -113,8 +113,11 @@ void HTMLVideoElement::set_video_track(JS::GCPtr<HTML::VideoTrack> video_track)
 void HTMLVideoElement::set_current_frame(Badge<VideoTrack>, RefPtr<Gfx::Bitmap> frame, double position)
 {
     m_current_frame = { move(frame), position };
-    if (paintable())
-        paintable()->set_needs_display();
+
+    if (auto* paintable = this->paintable()) {
+        paintable->set_needs_display();
+        paintable->document().invalidate_display_list();
+    }
 }

 void HTMLVideoElement::on_playing()
```

Having to call both `set_needs_display` and `invalidate_display_list` on each of these feels pretty clunk and easy to mess up 🤔

kalenikaliaksandr commented:
> Playing videos and interacting with the controls (e.g. volume) also update very slowly, if at all.

I guess there is a missing `invalidate_display_list()` for next video frame update?

> Having to call both set_needs_display and invalidate_display_list on each of these feels pretty clunk and easy to mess up 🤔

indeed, if something changes display list content and isn't a style/layout/scroll, then we have to call `invalidate_display_list`. agree, that it's clunky but I don't think there is some way better way to make this less error-prone. it's kinda the same as having to make sure layout is up to date in any function that relies on it.

trflynn89 commented:
>  guess there is a missing invalidate_display_list() for next video frame update?

Yep, that's covered in the patch I posted, as well as interacting with the controls on the painted media element.


Would it make sense to have `set_needs_display` also call `invalidate_display_list`? Is there is situation where we call the former and don't want to call the latter? 

gif animations are also broken for this same reason.

kalenikaliaksandr commented:
> Is there is situation where we call the former and don't want to call the latter?

yes, for example https://github.com/LadybirdBrowser/ladybird/blob/master/Userland/Libraries/LibWeb/HTML/Navigable.cpp#L2031 and https://github.com/LadybirdBrowser/ladybird/blob/master/Userland/Libraries/LibWeb/Painting/PaintableBox.cpp#L128 where we want to rasterize cached display list again with different scroll offsets.

> Would it make sense to have set_needs_display also call invalidate_display_list?

making them called together means no display list caching at all, because there won't be a way to trigger repainting without display list rebuilding.

tcl3 commented:
I've updated the PR to include the changes from the diff Tim posted above. I've also added commits to ensure that animated gifs work correctly and to ensure that the cursor is repainted when it is moved.

There are very likely other places that need to be updated, but these were places where I could easily see that the change was necessary.

It feels like we could have a separate method that calls `set_needs_display()` and `invalidate_display_list()` together. Naming suggestions welcome.

awesomekling commented:
> > Playing videos and interacting with the controls (e.g. volume) also update very slowly, if at all.
> 
> I guess there is a missing `invalidate_display_list()` for next video frame update?
> 
> > Having to call both set_needs_display and invalidate_display_list on each of these feels pretty clunk and easy to mess up 🤔
> 
> indeed, if something changes display list content and isn't a style/layout/scroll, then we have to call `invalidate_display_list`. agree, that it's clunky but I don't think there is some way better way to make this less error-prone. it's kinda the same as having to make sure layout is up to date in any function that relies on it.

Hmm, how about we combine them into a single call still though, something like

```cpp
set_needs_display(InvalidateDisplayList = InvalidateDisplayList::Yes);
```

Then you could override it with `::No` whenever you want to micromanage display list invalidation.

What do you think?

kalenikaliaksandr commented:
> > > Playing videos and interacting with the controls (e.g. volume) also update very slowly, if at all.
> > 
> > 
> > I guess there is a missing `invalidate_display_list()` for next video frame update?
> > > Having to call both set_needs_display and invalidate_display_list on each of these feels pretty clunk and easy to mess up 🤔
> > 
> > 
> > indeed, if something changes display list content and isn't a style/layout/scroll, then we have to call `invalidate_display_list`. agree, that it's clunky but I don't think there is some way better way to make this less error-prone. it's kinda the same as having to make sure layout is up to date in any function that relies on it.
> 
> Hmm, how about we combine them into a single call still though, something like
> 
> ```c++
> set_needs_display(InvalidateDisplayList = InvalidateDisplayList::Yes);
> ```
> 
> Then you could override it with `::No` whenever you want to micromanage display list invalidation.
> 
> What do you think?

that's a good idea! I think we can merge this PR and then I will do the change in `set_needs_display()`.

tcl3 commented:
I've now updated the PR so that `set_needs_display()` invalidates the display list by default, except when invalidating the display list is not necessary.

trflynn89 commented:
Do we need the first two commits if we have the third commit?

tcl3 commented:
> Do we need the first two commits if we have the third commit?

Oops, those weren't supposed to be there! Removed.
